staging/src/k8s.io/client-go/listers/generic_helpers.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package listers
+
+import (
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+)
+
+// ResourceIndexer wraps an indexer and resource for a given type.
+type ResourceIndexer[T runtime.Object] struct {
```

Comment:
API topic.  The way this is currently used is like:

```
listers.ResourceIndexer[*v1.FooBar]
```

In particular, the `*` is part of the type argument.  Could/should we simplfy that into:

```
listers.ResourceIndexer[v1.FooBar]
```

...which uses the actual type name, not the pointer.  Then signatures like:

```
func (l ResourceIndexer[T]) List(selector labels.Selector) (ret []*T, err error) {
  // ...
}

func (l ResourceIndexer[T]) Get(name string) (*T, error) {
  // ...
}
```

I personally find this more natural, but I have been known to get a bit too familiar with generics.

Reply:
That doesn’t work here because `T` is based on `runtime.Object`, and the latter’s methods all have pointer receivers, so `T` has to be a pointer type.